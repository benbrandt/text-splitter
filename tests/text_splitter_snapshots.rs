use std::fs;

use more_asserts::assert_le;
use once_cell::sync::Lazy;
use text_splitter::{Characters, ChunkSizer, TextSplitter};
use tiktoken_rs::{cl100k_base, CoreBPE};
use tokenizers::Tokenizer;

#[test]
fn characters_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default();
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(Characters.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn characters_range() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        let splitter = TextSplitter::default();
        let chunks = splitter.chunks(&text, 500..=2000).collect::<Vec<_>>();

        assert_eq!(chunks.join(""), text);
        for chunk in chunks.iter() {
            assert_le!(Characters.chunk_size(chunk), 2000);
        }
        insta::assert_yaml_snapshot!(chunks);
    });
}

#[test]
fn characters_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default().with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static HUGGINGFACE_TOKENIZER: Lazy<Tokenizer> =
    Lazy::new(|| Tokenizer::from_pretrained("bert-base-cased", None).unwrap());

#[test]
fn huggingface_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(HUGGINGFACE_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn huggingface_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static TIKTOKEN_TOKENIZER: Lazy<CoreBPE> = Lazy::new(|| cl100k_base().unwrap());

#[test]
fn tiktoken_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(TIKTOKEN_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn tiktoken_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}
