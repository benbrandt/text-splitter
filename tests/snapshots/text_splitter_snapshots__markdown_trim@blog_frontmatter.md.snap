---
source: tests/text_splitter_snapshots.rs
expression: chunks
input_file: tests/inputs/markdown/blog_frontmatter.md
---
- "---"
- "layout:"
- page
- "title:"
- Philosophy
- "permalink:"
- /
- philosophy
- "/\n---"
- Jekyll
- offers a
- unique
- philosophy
- when
- approachin
- g the
- problem of
- static
- site
- generation
- "."
- This core
- philosophy
- drives
- developmen
- t and
- product
- decisions.
- When a
- contributo
- "r,"
- maintainer
- ", or user"
- asks
- herself
- what
- Jekyll
- "is about,"
- the
- following
- principles
- should
- come to
- "mind:"
- "###"
- "1."
- No Magic
- Jekyll is
- not magic.
- A user
- should be
- able to
- understand
- the
- underlying
- processes
- that make
- up the
- Jekyll
- build
- without
- much
- reading.
- It should
- do only
- what you
- ask it to
- and
- nothing
- more.
- When a
- user takes
- a certain
- "action,"
- the
- outcome
- should be
- easily
- understand
- able and
- focused.
- "### 2. It"
- "\"Just"
- "Works\""
- The out-of
- "-the-box"
- experience
- should be
- "that it \""
- just works
- ".\" Run"
- "`gem"
- install
- "jekyll`"
- and it
- should
- build any
- Jekyll
- site that
- "it's given"
- "."
- Features
- like auto-
- regenerati
- on and
- settings
- like the
- markdown
- renderer
- should
- represent
- sane
- defaults
- that work
- perfectly
- for the
- vast
- majority
- of
- cases.
- The burden
- of initial
- configurat
- ion should
- not be
- placed on
- the user.
- "###"
- "3."
- Content is
- King
- Why is
- Jekyll so
- loved by
- content
- creators?
- It focuses
- on content
- first and
- "foremost,"
- making the
- process of
- publishing
- content on
- the Web
- easy.
- Users
- should
- find the
- management
- of their
- content
- enjoyable
- and simple
- "."
- "###"
- "4."
- Stability
- If a
- "user's"
- site
- builds
- "today, it"
- should
- build
- tomorrow.
- Backwards-
- compatibil
- ity should
- be
- strongly
- preferred
- over
- breaking
- changes.
- Breaking
- changes
- should be
- made to
- support a
- strong
- practical
- "goal, and"
- breaking
- changes
- should
- never be
- made to
- drive
- "forward \""
- "purity\" of"
- the
- "codebase,"
- or other
- changes
- purely to
- make the
- maintainer
- "s' lives"
- easier.
- Breaking
- changes
- provide a
- significan
- t amount
- of
- friction
- between
- upgrades
- and reduce
- the
- confidence
- of users
- in this
- "software,"
- and should
- thus be
- avoided
- unless
- absolutely
- necessary.
- Upon
- breaking
- "changes,"
- provide a
- clear path
- for users
- to upgrade
- "."
- "###"
- "5."
- Small &
- Extensible
- The core
- of Jekyll
- should be
- simple and
- "small, and"
- extensibil
- ity should
- be
- a first-
- class
- feature to
- provide
- added
- functional
- ity from
- community
- contributo
- rs.
- The core
- should be
- kept to
- features
- used by at
- least 90%
- of
- usersâ€“
- everything
- else
- should be
- provided
- as a
- plugin.
- New
- features
- should
- be shipped
- as plugins
- and focus
- should be
- put on
- creating
- extensible
- "core\nAPI'"
- s to
- support
- rich
- plugins.
